File: 1463.cpp
#include <math.h>

#include <iostream>
#include <vector>

using namespace std;

int ans(int);

vector<int> dp((int)pow(10, 6));

int main(void) {
  int num;
  cin >> num;
  cout << ans(num);
  dp[1] = 0;
}

int ans(int num) {
  for (int i = 2; i <= num; i++) {
    dp[i] = dp[i - 1] + 1;
    if (i % 2 == 0) dp[i] = min(dp[i - 1] + 1, dp[i / 2] + 1);
    if (i % 3 == 0) dp[i] = min(dp[i - 1] + 1, dp[i / 3] + 1);
  }
  return dp[num];
}

File: baekjoon/1463.py
from sys import stdin


def ans():
    dp = [0 for _ in range(N + 1)]

    dp[1] = 0

    for num in range(2, N + 1):
        if num % 6 == 0:
            dp[num] = min(dp[num // 2], dp[num // 3], dp[num - 1]) + 1
        elif num % 2 == 0:
            dp[num] = min(dp[num // 2], dp[num - 1]) + 1
        elif num % 3 == 0:
            dp[num] = min(dp[num // 3], dp[num - 1]) + 1
        else:
            dp[num] = dp[num - 1] + 1

    return dp[N]


N = int(stdin.readline())
print(ans())
